### PR TITLE
Query Passthrough with JDBC-Based Connectors Implementations

### PR DESCRIPTION
**Pull request by AbdulR3hman**
*Description of changes:*
This change will allow us to implement QPT (Query Pass-Through) for Athena Connectors. This is only a draft change to get early feedback on the SDK changes whilst making further improvement engine-side. 

Beside the changes to enable qpt SDK-side, I've also implemented QPT for MySQL, PostgreSQL, Redshift Connectors as a reference implementation.

*** Please Don't merge just yet ***

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/domain/Split.java [hunk 1]
```diff
@@ -239,6 +239,17 @@ public Builder withEncryptionKey(EncryptionKey key)
...
         }
 
+        /**
+         * Apply a map of properties to the split
+         * @param properties
+         * @return
+         */
+        public Builder applyProperties(Map<String, String> properties)
+        {
+            properties.entrySet().forEach(entry -> add(entry.getKey(), entry.getValue()));
```
Comment (chngpe) [inline athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/domain/Split.java:249]: can you rename properties from input to inputProperties or something. Right now the name is the same.

After that can we just this.properties.addAll(properties)

**Review by burhan94 — CHANGES_REQUESTED** [review]

File: athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/metadata/GetTableRequest.java [hunk 2]
```diff
@@ -23,18 +23,26 @@
...
 
+import java.util.Collections;
+import java.util.Map;
+
 import static java.util.Objects.requireNonNull;
 
 /**
  * Represents the input of a <code>GetTable</code> operation.
  */
+@JsonIgnoreProperties(value = {"queryPassthroughArguments"}, allowGetters = true, allowSetters = true)
```
[the remark this answers is not in the document]
Comment (AbdulR3hman) [reply athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/metadata/GetTableRequest.java:40]: this was added to allow for optional arguments; but with your previous feedback this won't be needed.

File: athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/handlers/MetadataHandler.java [hunk 3]
```diff
@@ -270,7 +272,7 @@ protected final void doHandleRequest(BlockAllocator allocator,
                 }
                 return;
             case GET_TABLE:
-                try (GetTableResponse response = doGetTable(allocator, (GetTableRequest) req)) {
+                try (GetTableResponse response = resolveDoGetTableImplementation(allocator, (GetTableRequest) req)) {
```
[the remark this answers is not in the document]
Comment (AbdulR3hman) [reply athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/handlers/MetadataHandler.java:275]: my reasoning is to allow the connector implementer to have a distinct interface that is needed to be implemented in order to allow to get QPT-based request; if this is over-kill happy to revert back to your previous (offline) recommendation?

File: athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/handlers/MetadataHandler.java [hunk 4]
```diff
@@ -522,4 +550,21 @@ private void assertTypes(GetTableResponse response)
             SupportedTypes.assertSupported(next);
         }
     }
+
+    /**
+     * Helper function that provides a single partition for Query Pass-Through
+     *
+     */
+    protected GetSplitsResponse setupQueryPassthroughSplit(GetSplitsRequest request)
```
[the remark this answers is not in the document]
Comment (AbdulR3hman) [reply athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/handlers/MetadataHandler.java:558]: I have moved this to the JDBC Metadata handler;

File: athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/domain/Split.java (hunk 1 above)
Comment (AbdulR3hman) [reply athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/domain/Split.java]: The SDK should be the only user of this interface; I don't think that would be a problem.

**Comment by AbdulR3hman** [comment]
closing this; will re-open a targeted PRs that are modular and focused thus easier to review.

**Pull request closed without merging**